### DISCUSSION TITLE
feat: impersistant package install (like `nix-shell -p`)

### DISCUSSION THREAD

**Discussion by aparatext** [Homebrew/discussions#4924] [Everyday usage]
First, Brew is one of the most excellent package managers/ecosystems I've seen! Packaging new apps is a actually pleasant, updates are shipped regularly, and if not, you can quickly rebuild it from HEAD with one flag, Brewfiles kind-of provide a declarative UX, and on top of that, everything is remarkably stable and installable everywhere without issues. (though on Linux, a few packages like Julia do break and I had to do `brew unlink util-linux` to make some apps happy)

However, I think Brew can borrow one powerful feature from Nix: one-time package install.

Say you need to run an application once, but not install it on the system. Maybe you need some tools CLI tools to run a script, maybe you're trying out an app and don't want to forget to remove it later, maybe it's an entire dev environment.

In Nix, you can use `nix-shell -p figlet cowsay lolcat` and it will drop you into a new shell with them installed and added to path (or the newer syntax similar to npx, `nix run figlet 'Hello!' -f smscript`). Once you're done, they get out of your path and will get removed next clean, or may be reused if invoked repeatedly (I may be misremembering, don't quote me on that).

I think having such a simple way to have the configuration stuff get out of the way temporarily, all the while leaving the actual persistant apps to the config, forcing you to still put diligence into maintaining your system, is a very interesting design choice I hope can be replicated in Brew.

I suppose you can do it with Brew currently with a Brewfile commit system, taking any untracked changes and resetting them on Cron while removing everything not tracked, but it's just my hack, seeing official support for it more in touch with Brew's vision would be much preferred.

Perhaps such a system can be extended to temporarily instantiate an isolated dev environment from Brewfile and destruct it when done?

**Comment by alebcay** [comment]
You may find [a previous exploration and discussion of such an idea (and its current implementation challenges and pitfalls)](https://github.com/Homebrew/brew/issues/11787) to be interesting.

The feature work has never gotten off the ground, but if you have something new to add to the discussion (w.r.t. the challenges presented in the thread - not much has improved on that front AFAIK) - not just in terms of desired usecases or user stories but rather how such a design could be implemented, it may be worth opening an issue or PR in `Homebrew/brew` for further consideration or discussion.
